mgr/dashboard: add RGW lifecycle management

Pegonzal opened the pull request:
Fixes: https://tracker.ceph.com/issues/50327


[screen-capture (1).webm](https://github.com/ceph/ceph/assets/106810838/a0f377c3-7939-4e9b-b735-0502c94c2aff)

<!--

  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

cloudbehl commented:
any updates?

Pegonzal commented:
> any updates?

I did update it last week, to make it work with xml or json, but I had some troubles with the parsing, so I still need to follow up on some changes

Pegonzal commented:
Currently adding json to xml parsing/checking from the backend. Also it is not possible to delete a lifecycleconfiguration yet, working on that as well. I'll try to add a few more tests after that

cloudbehl reviewed (commented):
Thanks Pedro.

1. Can we have toggle in the details page on Lifecycle section to see the JSON and XML both?
2. I think we should have scroll bar in Lifecycle section of details. As the policies and lifecycles can be huge data and other info we go out of the page.

Pegonzal commented:
jenkins test api

ivoalmeida commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:
```diff
@@ -424,18 +432,22 @@ export class RgwBucketFormComponent extends CdForm implements OnInit, AfterViewC
-    if (this.bucketPolicyTextArea) {
-      this.textAreaJsonFormatterService.format(this.bucketPolicyTextArea);
+  textAreaOnChange(textArea: ElementRef<any>) {
+    if (textArea && textArea.nativeElement?.value) {
+      if (textArea.nativeElement.value.startsWith('<')) {
+        this.textAreaXmlFormatterService.format(textArea);
+      } else {
+        this.textAreaJsonFormatterService.format(textArea);
+      }
     }
```
```
if (textArea && textArea.nativeElement?.value && textArea.nativeElement?.value?.startsWith?.('<')) {
  this.textAreaXmlFormatterService.format(textArea);
} else {
  this.textAreaJsonFormatterService.format(textArea);
}
    
```

cloudbehl reviewed (approved):
LGTM. Tested and things are working

Minor comments
1.https://github.com/ceph/ceph/pull/57238/files#r1625457453
2. If we don't have any lifecycle policy set we can hide the toggle and the default value can be `{}` like we have for bucket policies.

Pegonzal commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:
```diff
@@ -468,6 +468,50 @@
+              <label i18n
+                     class="cd-col-form-label"
+                     for="id">Lifecycle
+                <cd-helper>JSON or XML formatted document</cd-helper>
```
I personally though this was non essential information and since there also is the validator error and text message, I left it as a helper.

But If you would prefer it to have it upfront we can change it. I'm not sure where should lay this text, where would you place the text @cloudbehl? So I can modify it accordingly

Pegonzal commented:
Addressed the comments.

Also changed:
```
-       if replication is not None:
+       if replication:
```

To fix following error when editing bucket:
```
2024-06-26T19:08:43.423+0000 7f317ab6a700  0 [dashboard INFO exception] Dashboard Exception: RGW REST API failed request with status code 405
(b'{"Code":"MethodNotAllowed","Message":"","RequestId":"tx0000075c8c08e469255c0'
 b'-00667c673b-4971-zone1-zg1-realm1","HostId":"4971-zone1-zg1-realm1-zg1-realm'
 b'1"}')

```
from #57930 also seen [here](https://github.com/ceph/ceph/pull/57930#issuecomment-2182433187):

nizamial09 commented:
> from #57930 also seen [here](https://github.com/ceph/ceph/pull/57930#issuecomment-2182433187)

btw, that was a different error which happened because of improper daemon configuration (which was from one of my other commits in the PR)

nizamial09 commented:
make check failure
```
************* Module controllers.rgw
controllers/rgw.py:536:4: R0912: Too many branches (13/12) (too-many-branches)
```

I'd advice, before committing try to run the linters locally so that can be caught and dealt with sooner than waiting for the jenkins.
